python/tests/test_pyvw.py

```diff
@@ -347,94 +381,158 @@ def check_error_raises(type, argument):
     with pytest.raises(type) as error:
         argument()
 
+
 def test_from_colnames_constructor():
     df = pd.DataFrame({"y": [1], "x": [2]})
     conv = DFtoVW.from_colnames(y="y", x=["x"], df=df)
-    lines_list = conv.process_df()
+    lines_list = conv.convert_df()
     first_line = lines_list[0]
     assert first_line == "1 | 2"
```

Comment:
I just noticed that there isn't a test that verifies past one line, can you modify one of the tests and an assert on the second_line just to have more coverage for future PRs.

Reply:
also, would there be any value on having a test of a big DF being converted? Maybe not in this file, but for CI purposes

Reply:
@lalo 

Regarding your **1st** comment : a test with multiple lines is performed line 428 of the _new_ file (`test_multiple_lines`).

Regarding your **2nd** comment : of course, a test to see how it scales could be included.

I personally tested the class on [avazu click through rate prediction dataset from kaggle](https://www.kaggle.com/c/avazu-ctr-prediction/data). The class takes about 10 seconds to convert 1 millions rows with about 10 columns. The **main bottleneck is the conversion of float/int columns to string columns** (almost 1 sec for 1 million rows). You can see the related question I post on [StackOverflow](https://stackoverflow.com/questions/61445646/concatenate-the-values-of-two-pandas-series-that-are-not-of-str-type-is-too-sl). 

I can add a test using random data with 100k, 1million or more rows.

Reply:
There's 2 things at play here
1) we shouldn't be converting into the string to send it off to vw but this is just room for improvement on the python api side, not related to vw. So in the long term, there might be performance gains.
2) I wonder if we can improve performance by switching things up in the functions that create the string. For example if we use df.iterrows to iterate over each row and create the line on one go, instead of accessing each value from different places. Not sure if there would be a benefit, though. Another way would be by converting the df to a numpy.array and operate directly from those values. This might come into play if we had what i mentioned on the previous point available.

Reply:
1. If not string, what should format sent ?

2. Iterate through the rows takes much more times, the main benefit of using columns operation is we take advantage of vectorization. Also converting to `numpy.array` won't result in huge gains since it's what pandas already use in the backend. Indeed you can try the following in ipython console : 

```python
import pandas as pd
import numpy as np

x = np.random.normal(size=int(1e6))
s = pd.Series(x)

%timeit s.astype(str) # 1.42 s ± 79.6 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
%timeit  s.to_numpy().astype(str) # 1.27 s ± 83.8 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
```

Reply:
1) Binary data and try to avoid converting to strings just to pass it down elsewhere
2) Good point on the column. I was going for minimizing string concat operations and building the string in one go. Anyways, its best to measure before going on possible dead ends.

Could you share the python script to load the avazu dataset? Or maybe add it as a python example (there is a dir for these scripts) if you want to

Reply:
1. Okay I see
2. Yes when I did the original PR 1 month ago I tried various methods too (line per line, converting to numpy and doing ops in numpy, doing operation directly on pandas columns).

Of course I can share the script to load the avazu dataset ! You examples foldder is the following https://github.com/VowpalWabbit/vowpal_wabbit/tree/master/python/examples ? 

Please note that the dataset is huge (the training part is about 44 millions lines and 6 Gib uncompressed), so not sure if this is to be uploaded :smile: I can do sampling too. As the dataset does not fit in memory I used pandas generator to load it in batch of 1 million rows and learn on each batch. I think it can be a good showcase on how to use the DFtoVW API and to learn on data that wouldn't fit in memory (within Python session).

Reply:
Yes, that's the folder. I think we could just reference the dataset on a python comment on the script. It would still be helpful to see an actual example of using DFtoVW. Later, if someone wants to run it they have to download the dataset themselves. Feel free to open another PR.